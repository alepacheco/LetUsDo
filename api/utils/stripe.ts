import * as Stripe from 'stripe';

type HandleErrorArgs = {
    status?: string;
    next_action?: {
        type: string;
    };
    client_secret?: string;
};

export const handleError = async (func: () => Promise<HandleErrorArgs>) => {
    try {
        const { status, next_action, client_secret } = await func();

        if (status === 'requires_action' && next_action && next_action.type === 'use_stripe_sdk') {
            return {
                status: 200,
                response: {
                    requires_action: true,
                    payment_intent_client_secret: client_secret
                }
            };
        }
        if (status === 'succeeded') {
            return {
                status: 200,
                response: {
                    success: true
                }
            };
        }
    } catch (error) {
        return {
            status: 500,
            response: {
                error
            }
        };
    }

    return {
        status: 500,
        error: 'Unknown error while executing payment'
    };
};

type ExecutePaymentMethod = {
    payment_method_id: string;
    amount: number;
    taskText: string;
    metadata: {
        email: string;
        remoteAddress: string;
        name: string;
        phone: string;
    };

    stripe: Stripe;
};
export const executePaymentMethod = async ({
    payment_method_id,
    amount,
    taskText,
    metadata,
    stripe
}: ExecutePaymentMethod) => handleError(() => stripe.paymentIntents.create({
    payment_method: payment_method_id,
    amount,
    description: `Task details: ${taskText}`,
    currency: 'gbp',
    receipt_email: metadata.email,
    confirmation_method: 'manual',
    confirm: true,
    metadata
}));


type ExecutePaymentIntent = {
    payment_intent_id: string;
    stripe: Stripe;
};
export const executePaymentIntent = async ({ stripe, payment_intent_id }: ExecutePaymentIntent) =>
    handleError(() => stripe.paymentIntents.confirm(
        payment_intent_id
    ));
