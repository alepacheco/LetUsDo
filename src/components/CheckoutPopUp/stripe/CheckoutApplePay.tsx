import React from 'react';
import { injectStripe, PaymentRequestButtonElement } from 'react-stripe-elements';
import * as actions from 'src/actions/taskModalActions';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

class CheckoutApplePay extends React.Component<any, any> {
  constructor(props: any) {
    super(props);

    const paymentRequest = props.stripe.paymentRequest({
      country: 'GB',
      currency: 'gbp',
      total: {
        label: 'Task fixed price',
        amount: 99
      },
      requestPayerName: true,
      requestPayerEmail: true,
      requestPayerPhone: true
    });

    paymentRequest.on('token', ({ complete, token, ...data }: any) => {
      console.log('Received Stripe token: ', token);
      console.log('Received customer information: ', data);
      complete('success');
    });

    paymentRequest.canMakePayment().then((result: any) => {
      this.setState({ canMakePayment: !!result });
      this.props.actions.setApplePayAvailable(!!result);

      console.log({ canMakePayment: !!result });
    });

    this.state = {
      canMakePayment: false,
      paymentRequest
    };
  }

  render() {
    return this.state.canMakePayment ? (
      <PaymentRequestButtonElement
        paymentRequest={this.state.paymentRequest}
        className="PaymentRequestButton"
      />
    ) : null;
  }
}

const mapDispatchToProps = (dispatch: any) => ({
  actions: bindActionCreators(actions, dispatch)
});

export default connect(
  () => ({}),
  mapDispatchToProps
)(injectStripe(CheckoutApplePay));
