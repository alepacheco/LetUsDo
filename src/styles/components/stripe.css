.card-numer-field {
    border: 2px solid #dee2e6;
    border-radius: 2px;
    padding: 8px;
    margin: 8px 0;
}

.pay-button-wrapper {
    text-align: right;
}

.pay-button {

}