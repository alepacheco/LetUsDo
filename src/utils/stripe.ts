import axios from 'axios';
import { ReactStripeElements } from 'react-stripe-elements';

type TryPayment = {
  taskText: string;
  email?: string;
  name?: string;
  phone?: string;
  payment_method_id?: string;
  payment_intent_id?: string;
  stripe: ReactStripeElements.StripeProps;
};

export const tryPayment = async ({
  taskText,
  email,
  name,
  phone,
  payment_method_id,
  payment_intent_id,
  stripe
}: TryPayment): Promise<boolean | undefined> => {
  const {
    data: { requires_action, payment_intent_client_secret, success, error }
  } = await axios.post('/api/createPayment', {
    payment_method_id,
    payment_intent_id,
    taskText,
    metadata: {
      email,
      name,
      phone,
    }
  });

  if (error) {
    throw new Error(error);
  }

  if (success) {
    return true;
  }

  if (requires_action) {
    // @ts-ignore
    const { error, paymentIntent } = await stripe.handleCardAction(payment_intent_client_secret);

    if (error || !paymentIntent.id) {
      throw error;
    }

    return tryPayment({
      payment_intent_id: paymentIntent.id,
      taskText,
      email,
      name,
      phone,
      stripe
    });
  }
};

type SubmitProps = {
  email: string;
  stripe: ReactStripeElements.StripeProps;
  taskText: string;
};
export const submitPayment = async ({ email, stripe, taskText }: SubmitProps): Promise<boolean> => {
  try {
    const { paymentMethod, error } = await stripe.createPaymentMethod('card');

    if (error || !paymentMethod || !paymentMethod.id) {
      throw error;
    }

    return (
      (await tryPayment({
        taskText,
        email,
        payment_method_id: paymentMethod.id,
        stripe
      })) || false
    );
  } catch (error) {
    console.log(error);
    return false;
  }
};
