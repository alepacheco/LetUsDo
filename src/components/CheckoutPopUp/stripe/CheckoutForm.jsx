/* eslint-disable no-console */
import React from 'react';
import { bindActionCreators } from 'redux';
import Button from 'react-bootstrap/Button';
import Badge from 'react-bootstrap/Badge';
import { connect } from 'react-redux';
import { CardElement, injectStripe } from 'react-stripe-elements';
import '../../../styles/components/stripe.css';
import PropTypes from 'prop-types';
import axios from 'axios';
import * as actions from '../../../actions/taskModalActions';

export class CheckoutForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = { complete: false };
    this.submit = this.submit.bind(this);
  }

  async submit() {
    const { token, error: createTokenError } = await this.props.stripe.createToken({
      name: 'Let Us Do'
    });
    if (!token) {
      console.log(createTokenError);
      return;
    }

    try {
      await axios.post('/api/createPayment', {
        token,
        email: this.props.email,
        taskText: this.props.taskText
      });

      this.setState({ complete: true });

      this.props.actions.setDialog('purchaseCompleted');
    } catch (error) {
      console.log(error.message);
    }
  }

  render() {
    // TODO improve
    if (this.state.complete) {
      return <h1>Purchase Complete, we will get in touch with updates to your task</h1>;
    }

    return (
      <div className="checkout">
        <div className="card-numer-field">
          <CardElement />
        </div>
        <div className="pay-button-wrapper">
          <Button className="pay-button" onClick={this.submit} disabled={!this.props.validEmail}>
            Pay{' '}
            <Badge pill variant="light">
              20Â£
            </Badge>
          </Button>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  validEmail: state.checkoutPopUp.validEmail,
  email: state.checkoutPopUp.email,
  taskText: state.taskModal.taskText
});
function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators(actions, dispatch)
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(injectStripe(CheckoutForm));

CheckoutForm.propTypes = {
  validEmail: PropTypes.bool,
  email: PropTypes.string,
  taskText: PropTypes.string,
  stripe: PropTypes.object,
  actions: PropTypes.shape({
    setDialog: PropTypes.func
  }).isRequired
};
