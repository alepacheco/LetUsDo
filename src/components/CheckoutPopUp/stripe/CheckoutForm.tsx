/* eslint-disable no-console */
import React from 'react';
import { bindActionCreators } from 'redux';
import Button from 'react-bootstrap/Button';
import Badge from 'react-bootstrap/Badge';
import { connect } from 'react-redux';
import { CardElement, injectStripe, ReactStripeElements } from 'react-stripe-elements';
import { trackEvent } from 'src/utils/analytics';
import 'src/styles/components/stripe.css';
import { submitPayment } from 'src/utils/stripe';
// @ts-ignore
import * as actions from 'src/actions/taskModalActions';

const tryApplePay = async ({ stripe }: { stripe: ReactStripeElements.StripeProps }) => {
  const paymentRequest = stripe.paymentRequest({
    country: 'GB', // TODO check
    currency: 'gbp',
    total: {
      label: 'Price',
      amount: 50
    },
    requestPayerName: true,
    requestPayerEmail: true
  });

  const canUsePayBotton = await paymentRequest.canMakePayment();
  if (canUsePayBotton) {
    // @ts-ignore
    const elements = stripe.elements();

    const prButton = elements.create('paymentRequestButton', {
      paymentRequest
    });

    prButton.mount('#checkout-bottom');

    const element = document.getElementById('checkout-form');
    if (element) {
      element.style.display = 'none';
    }
  } else {
    const element = document.getElementById('checkout-bottom');
    if (element) {
      element.style.display = 'none';
    }
    return;
  }

  // @ts-ignore
  const elements = stripe.elements();

  const prButton = elements.create('paymentRequestButton', {
    paymentRequest
  });

  prButton.mount('#checkout-bottom');

  const element = document.getElementById('checkout-form');
  if (element) {
    element.style.display = 'none';
  }

  paymentRequest.on('paymentmethod', async ev => {
    // @ts-ignore
    const { error: confirmError, paymentIntent } = await stripe.confirmPaymentIntent(
      process.env.STRIPE_FRONT,
      {
        payment_method: ev.paymentMethod.id
      }
    );

    if (confirmError || !process.env.STRIPE_FRONT) {
      // Report to the browser that the payment failed, prompting it to
      // re-show the payment interface, or show an error message and close
      // the payment interface.
      console.log('apple pay failed');
      ev.complete('fail');
    } else {
      // Report to the browser that the confirmation was successful, prompting
      // it to close the browser payment method collection interface.
      console.log('apple pay collected');
      ev.complete('success');
      // Let Stripe.js handle the rest of the payment flow.
      const { error } = await stripe.handleCardPayment(process.env.STRIPE_FRONT);
      if (error) {
        console.log('apple pay failed');
        // The payment failed -- ask your customer for a new payment method.
      } else {
        console.log('apple pay payed');
        // The payment has succeeded.
      }
    }
  });
};

export const CheckoutForm: React.FC<{
  email: string;
  taskText: string;
  stripe?: ReactStripeElements.StripeProps;
  validEmail: boolean;
  actions: { setDialog: (state: string) => void };
}> = ({ email, stripe, taskText, actions: { setDialog }, validEmail }) => {
  if (stripe) {
    // tryApplePay({ stripe });
  }

  const onClick = async () => {
    console.log('Set loading here but keep the element alive.');
    trackEvent({
      category: 'click',
      action: 'pay button clicked'
    });
    trackEvent({
      category: 'purchase',
      action: 'purchase initiated'
    });

    if (!stripe) {
      setDialog('purchaseFailed');

      return;
    }

    const completed = await submitPayment({
      email,
      stripe,
      taskText
    });
    if (completed) {
      setDialog('purchaseCompleted');
      trackEvent({
        category: 'purchase',
        action: 'purchase completed'
      });
    } else {
      setDialog('purchaseFailed');
      trackEvent({
        category: 'purchase',
        action: 'purchase failed'
      });
    }
  };

  return (
    <div className="checkout">
      <div id="checkout-bottom" />
      <div id="checkout-form">
        <div className="card-numer-field">
          <CardElement />
        </div>
        <div className="pay-button-wrapper">
          <Button className="pay-button" onClick={onClick} disabled={!validEmail}>
            Pay{' '}
            <Badge pill variant="light">
              20Â£
            </Badge>
          </Button>
        </div>
      </div>
    </div>
  );
};

const mapStateToProps = (state: any) => ({
  validEmail: state.checkoutPopUp.validEmail,
  email: state.checkoutPopUp.email,
  taskText: state.taskModal.taskText
});
const mapDispatchToProps = (dispatch: any) => ({
  actions: bindActionCreators(actions, dispatch)
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(injectStripe(CheckoutForm));
