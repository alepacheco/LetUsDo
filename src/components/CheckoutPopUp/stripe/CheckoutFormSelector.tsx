/* eslint-disable no-console */
import React from 'react';
import 'src/styles/components/stripe.scss';
import { connect } from 'react-redux';
import CheckoutForm from './CheckoutForm';
import CheckoutApplePay from './CheckoutApplePay';

export const CheckoutFormSelector: React.FC<{ applePayAvailable: boolean; loading: boolean }> = ({
  applePayAvailable,
  loading
}) => {
  if (applePayAvailable || applePayAvailable === null) {
    return <CheckoutApplePay />;
  }
  return <CheckoutForm loading={loading} />;
};

const mapStateToProps = (state: any) => ({
  applePayAvailable: state.taskModal.applePayAvailable
});

export default connect(mapStateToProps)(CheckoutFormSelector);
