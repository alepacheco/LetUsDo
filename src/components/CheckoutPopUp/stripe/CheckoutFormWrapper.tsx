import React from 'react';
import { Elements, StripeProvider } from 'react-stripe-elements';
import CheckoutForm from './CheckoutForm';

export const CheckoutFormWrapper = () => (
  // @ts-ignore
  <StripeProvider apiKey={process.env.STRIPE_FRONT}>
    <Elements>
      <CheckoutForm />
    </Elements>
  </StripeProvider>
);

export default CheckoutFormWrapper;
