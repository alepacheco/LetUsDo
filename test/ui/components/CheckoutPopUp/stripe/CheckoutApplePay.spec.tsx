/* eslint-disable no-undef */
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { CheckoutApplePay } from 'components/CheckoutPopUp/stripe/CheckoutApplePay';

const flushPromises = () => new Promise(setImmediate);

describe('<CheckoutApplePay />', () => {
  const stripeMock: any = {
    paymentRequest: () => ({
      on: () => ({}),
      show: () => ({}),
      canMakePayment: async () => true
    })
  };
  const props = {
    setApplePayAvailable: true,
    actions: {
      setApplePayAvailable: jest.fn()
    },
    stripe: stripeMock,
    taskText: 'Task description'
  };

  it('renders', async () => {
    const wrapper = shallow(<CheckoutApplePay {...props} />);
    await flushPromises();

    expect(toJson(wrapper)).toMatchSnapshot();
    expect(props.actions.setApplePayAvailable).toHaveBeenLastCalledWith(true);
  });

  it('not renders if disabled', async () => {
    const stripeMock: any = {
      paymentRequest: () => ({
        on: () => ({}),
        show: () => ({}),
        canMakePayment: async () => false
      })
    };
    const props = {
      setApplePayAvailable: false,
      actions: {
        setApplePayAvailable: jest.fn()
      },
      stripe: stripeMock,
      taskText: 'Task description'
    };

    const wrapper = shallow(<CheckoutApplePay {...props} />);
    await flushPromises();

    expect(toJson(wrapper)).toMatchSnapshot();
    expect(props.actions.setApplePayAvailable).toHaveBeenLastCalledWith(false);
  });
});
