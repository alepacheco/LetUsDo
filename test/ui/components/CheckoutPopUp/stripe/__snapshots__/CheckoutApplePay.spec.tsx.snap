// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CheckoutApplePay /> not renders if disabled 1`] = `""`;

exports[`<CheckoutApplePay /> renders 1`] = `
<PaymentRequestButtonElement
  className="PaymentRequestButton"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onReady={[Function]}
  paymentRequest={
    Object {
      "canMakePayment": [Function],
      "on": [Function],
      "show": [Function],
    }
  }
/>
`;
